File: 0463-island-perimeter/0463-island-perimeter.java
class Solution {
    HashSet<String> seen = new HashSet<>();
    public int islandPerimeter(int[][] grid) {
        for (int i=0; i < grid.length; i++){
            for (int j=0; j < grid[0].length; j++){
                if (grid[i][j] == 1) return dfs(i, j, grid);
            }
        }
        return 0;
    }
    
    public int dfs(int i, int j, int[][] grid){
        if (i<0 || j<0 || i >= grid.length || j >= grid[0].length || grid[i][j] == 0) return 1;
        String temp = i + "," + j;
        
        if (seen.contains(temp)) return 0;
        else seen.add(temp);
        
        int ans = 0;
        
        ans += dfs(i+1, j, grid);
        ans += dfs(i-1, j, grid);
        ans += dfs(i, j+1, grid);
        ans += dfs(i, j-1, grid);
        
        return ans;
    }
}

File: 0463-island-perimeter/0463-island-perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        seen = set()
        
        def dfs(i, j):
            if i >= len(grid) or i < 0 or j >= len(grid[0]) or j < 0 or grid[i][j] == 0:
                return 1
            
            if (i,j) in seen:
                return 0
            
            seen.add((i,j))
            ans = dfs(i, j+1)
            ans += dfs(i, j-1)
            ans += dfs(i+1, j)
            ans += dfs(i-1, j)
            return ans
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j]:
                    return dfs(i,j)